### ISSUE TITLE
Gradeable categories cannot gain focus

### ISSUE THREAD

**Issue by Kevin-MM** [Submitty/Submitty#4117]
On the gradeables page the category headings cannot gain focus.  For example these would be ```FUTURE    visible only to Instructors``` or ```OPEN``` or ```PAST DUE``` when looking at the sample course.

Since these items are skipped over when tabbing, this information is basically lost to screen reader only users.

**Comment by MasterOdin** [comment]
I'm not really sure what this means? These headers aren't links or have any action to them so why would they be accessed via tabbing, or that the lack of being able to explicitly tab to a given piece of content on the page means that it's lost to screen readers as that would mean pretty much all content on every page is lost to screen readers?

**Comment by Kevin-MM** [comment]
For the screen reader to read an element, that element must be able to gain focus (using the tab button, you can stop on that element).  Imagine being a user that does not use any type of accessibility software.  You open up the gradeables page, and you can clearly see that the gradeables are broken into categories.  You know within a few seconds of looking at the page which gradeables are open, which ones are past due, etc.  Now imagine being a blind user, the only information you learn about the page is that which the screen reader audibly reads to you.  You press tab, but the screen readers skips right over these headings, this is what I mean when I say that information is basically lost to screen reader only users.  Those users have absolutely no way of knowing that the gradeables are broken into convenient categories.

**Comment by MasterOdin** [comment]
Go to https://webaim.org/techniques/keyboard/ (or any page on the site) and hit tab, see what elements you land on. 

See also this quote from the page:
> ### Focus indicators
> A keyboard user typically uses the Tab key to navigate through interactive elements on a web page—links, buttons, fields for inputting text, etc. When an item has keyboard "focus", it can be activated or manipulated with the keyboard.

Given that the headers don't have any keyboard action associated with them, not sure why the elements would get focus via tab, and that from my understanding, screen readers can go through a page in several various ways so long as the structure of the page is good and proper.

**Comment by Kevin-MM** [comment]
Point taken, although I think @pbailie needs to review this issue to shed some more light on best practices here.

**Comment by pbailie** [comment]
I haven't seen (heard) macOS's built in assistive screen reader actually say anything that doesn't get tab focus — in Chrome, at least.  But I really want to see how a big-brand screen reader behaves.  I recently borrowed a (not so great) laptop with Kerzweil installed.  I'll get back to you.

**Comment by pbailie** [comment]
This is just an initial observation, but what I am observing is that Kurzweil will read webpages in the DOM order (it even dictating "skip to main content" even though it wasn't visible), but does not pay any attention to TAB focus.  It just keeps on reading/dictating.  On the plus side, it will (eventually) read everything, including elements that cannot get focus.  On the other hand, it pays no attention to focus, so it isn't useful as a guide to the UI.

I am going to research this further.  It has to be clunky to utilize both Kurzweil (to read the whole page) and another screen reader (for UI interaction).

**Comment by pbailie** [comment]
Update:  I did manage to get Apple Voiceover (the screen reader built into macOS) to navigate through text only elements.  `TAB` is meant to navigate among links quickly.  However, other keyboard shortcuts will allow the screen reader to "focus" on non-interactive portions.

e.g. In Voice over, `CRTL+OPTION` + `down-arrow` allowed me to "drop down" into more of the page at Heading level 1.  Then `CTRL+OPTION` +`left-arrow` or `right-arrow` allowed me to skip around and read non-interactive text blocks.  In fact, this portion showed just how _critical_ it is that a page has a proper heading outline, proper DOM order of elements, proper designation of data and layout for tables, etc.

But really important, here, is that screen readers are not intuitive to most of us.  And that actually makes sense considering assistive technologies are designed to provide _different_ user functionality.  This is a good thing for those who need it.  That also means that to test a webpage with a screen reader will require patience as it will mean learning to use something very unfamiliar.

But to answer the original question:  Screen readers can get to content that `TAB` skips over.  It just requires a different keyboard shortcut, which varies from reader to reader.  e.g. Using Apple Voice Over, `CTRL+OPTION` + `left-arrow` or `right-arrow` will help get around more.

If you are using Windows, I have read recommendations towards using NVDA, which is free and (mostly) open source.  I have not yet discovered its navigation controls.  [https://www.nvaccess.org/download/](https://www.nvaccess.org/download/)

I intend to create a more in depth guide in using these technologies — probably focused on Voice Over and NVDA.  I may not get to it right away as I do have a few other things get done.

I am closing the issue as it is possible for screen readers to access content that `TAB` skips over.

**Comment by Kevin-MM** [comment]
I found this for ChromeVox incase anyone is interested in testing

http://www.chromevox.com/navigating.html
